src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/utils/RangeUtils.java

```diff
@@ -0,0 +1,104 @@
+package uk.gov.homeoffice.digital.sas.balancecalculator.utils;
+
+import com.google.common.collect.Range;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.LongStream;
+
+public class RangeUtils {
+
+  private RangeUtils() {
+
+  }
+
+  private static final ZoneId UK_TIME_ZONE = ZoneId.of("Europe/London");
+
+  public static SortedMap<LocalDate, Range<ZonedDateTime>> splitOverDays(
+                                                                      ZonedDateTime startDateTime,
+                                                                      ZonedDateTime endDateTime) {
+
+    SortedMap<LocalDate, Range<ZonedDateTime>> intervals = new TreeMap<>();
+
+    // The split should be based on UK time
+    ZonedDateTime ukStartTime = startDateTime.withZoneSameInstant(UK_TIME_ZONE);
+    ZonedDateTime ukEndTime = endDateTime.withZoneSameInstant(UK_TIME_ZONE);
+
+    var numDaysCovered = getNumOfDaysCoveredByDateRange(ukStartTime, ukEndTime);
+    if (numDaysCovered == 1) {
+      intervals.put(ukStartTime.toLocalDate(),
+          Range.closed(ukStartTime, ukEndTime));
+    } else {
+
+      intervals.put(ukStartTime.toLocalDate(), RangeUtils.startDayRange(ukStartTime));
+
+      if (numDaysCovered > 2) {
+        intervals.putAll(RangeUtils.midDayRangesMap(ukStartTime.plusDays(1),
+            numDaysCovered - 1));
+      }
+
+      if (RangeUtils.endDayRange(ukEndTime) != null) {
+        intervals.put(ukEndTime.toLocalDate(), RangeUtils.endDayRange(ukEndTime));
+      }
+    }
+
+    return intervals;
+  }
+
+  private static Range<ZonedDateTime> startDayRange(ZonedDateTime startDateTime) {
+
+    return Range.closed(
+        startDateTime,
+        ZonedDateTime.of(
+            startDateTime.plusDays(1L).toLocalDate().atTime(0, 0),
+            startDateTime.getZone()
+        ));
+  }
+
+  private static Range<ZonedDateTime> endDayRange(ZonedDateTime endDateTime) {
+
+    Range<ZonedDateTime> endDayRange = Range.closed(
+        ZonedDateTime.of(endDateTime.toLocalDate().atTime(0, 0),
+            endDateTime.getZone()),
+        endDateTime
+    );
+    return Duration.between(endDayRange.lowerEndpoint(),
+        endDayRange.upperEndpoint()).toMinutes() >= 1
+        ? endDayRange : null;
```

Comment:
Should this logic be its own function rather than returning null?

Reply:
Sure. Fixed